bin/reflective_loader/src/loader.rs

```diff
@@ -0,0 +1,836 @@
+use core::ptr;
+use core::ffi::c_void;
+
+ // This is an assumption that may be wrong. Due to no_std it's ideal to keep things
+ // on the stack which has a default size of 1024KB. If more sections are needed
+ // we can rework this hopefully by then a no_std zero-copy pe parser will exist
+ // and our home grown PE parser can be removed.
+const MAX_PE_SECTIONS: usize = 32;
+const PE_MAGIC: u16 = 0x5A4D; // MZ
+const NT_SIGNATURE: u32 = 0x4550; // PE
+
+
+const KERNEL32_HASH: u32 = 0x6ddb9555;
+const NTDLL_HASH: u32 = 0x1edab0ed;
+const LOAD_LIBRARY_A_HASH: u32 = 0xb7072fdb;
+const GET_PROC_ADDRESS_HASH: u32 = 0xdecfc1bf;
+const VIRTUAL_ALLOC_HASH: u32 = 0x97bc257;
+const GET_LAST_ERROR_HASH: u32  = 0x8160BDC3;
+
+#[derive(Copy,Clone)]
+struct UserData {
+    function_offset: u64,
+}
+
+type FnDllMain = unsafe extern "system" fn(module: HINSTANCE, call_reason: u32, reserved: *mut c_void) -> BOOL;
+
+#[allow(non_camel_case_types)]
+type generic_fn = unsafe extern "system" fn() -> ();
+
+type FnLoadLibraryA = unsafe extern "system" fn(lplibfilename: PCSTR) -> HINSTANCE;
+
+type FnGetProcAddress = unsafe extern "system" fn(hmodule: HINSTANCE, lpprocname: PCSTR) -> FARPROC;
+
+type FnVirtualAlloc = unsafe extern "system" fn(lpaddress: *const c_void, dwsize: usize, flallocationtype: VIRTUAL_ALLOCATION_TYPE, flprotect: PAGE_PROTECTION_FLAGS) -> *mut c_void;
+
+type FnGetLastError = unsafe extern "system" fn() -> u32;
+
+// pub unsafe fn VirtualAlloc(hprocess: super::super::Foundation::HANDLE, lpaddress: *const ::core::ffi::c_void, dwsize: usize, flallocationtype: VIRTUAL_ALLOCATION_TYPE, flprotect: PAGE_PROTECTION_FLAGS) -> *mut ::core::ffi::c_void
+fn virtual_alloc(fn_ptr: FnVirtualAlloc, err_ptr: FnGetLastError, lp_address: *const c_void, dw_size: usize, fl_allocation_type: u32, fl_protect: u32) -> *mut c_void {
+    let buffer_handle: *mut c_void = unsafe{ fn_ptr(lp_address, dw_size, fl_allocation_type, fl_protect) };
+    if buffer_handle == null_mut() {
+        let error_code = unsafe { err_ptr() };
+        if error_code != 0 {
+            panic!("Failed to allocate memory. Last error returned: {}", error_code);
+        }
+    }
+    buffer_handle
+}
+
+// pub unsafe fn GetProcAddress(hmodule: P0, lpprocname: P1) -> FARPROC
+fn get_proc_address(fn_ptr: FnGetProcAddress, err_ptr: FnGetLastError, h_module: isize, lp_proc_name: *const u8) -> isize {
+    let proc_handle = match unsafe { fn_ptr(h_module, lp_proc_name) } {
+        Some(local_proc_handle) => local_proc_handle,
+        None => {
+            let error_code = unsafe { err_ptr() };
+            panic!("Failed to find function {:?} in module. Last error returned: {}", lp_proc_name, error_code);
+        }
+    };
+    proc_handle as isize
+}
+
+
+// pub unsafe fn LoadLibraryA(lplibfilename: P0) ->  Result<HMODULE>
+fn load_library_a(fn_ptr: FnLoadLibraryA, err_ptr: FnGetLastError, lplibfilename: PCSTR) -> HINSTANCE {
```

Comment:
For your consideration, some witch magic used in libloading make syms callable via deref. 
https://docs.rs/libloading/latest/src/libloading/os/windows/mod.rs.html#368